### PR TITLE
[FLINK-25176][table] Introduce "ALTER TABLE ... COMPACT" syntax

### PR DESCRIPTION
**Pull request by LadyForest**
## What is the purpose of the change

This PR is a foundation of FLINK-25176 that introduces the SQL syntax `ALTET TABLE table_identifier [PARTITION partition_spec] COMPACT`, which paves the way for coordination with managed table storage. You can find more details at [FLIP-188](https://cwiki.apache.org/confluence/display/FLINK/FLIP-188%3A+Introduce+Built-in+Dynamic+Table+Storage).

## Brief changelog
This PR contains two commits
- 69a521de
   - fix typo in `UseCatalogOperation`
   - fix incorrect `{@link}` in `PartitioningSpec` (which I've confirmed with @godfreyhe)
   - unify the `asSummaryString` method in `AlterTableDropConstraintOperation` to use `tableIdentifier.asSummaryString` because `tableIdentifier.toStirng` is implicitly call `tableIdentifier.asSerializableString`
   - migrate all tests in`SqlToOperationConverterTest` to assertJ

- 980177f1
  - Extend `Parser.tdd` and `parserImpls.ftl` templates for flink-sql-parser  to support syntax. Add `SqlAlterTableCompact` as the representition after parsing.
  - Extend `SqlToOperationConverter` to support operation transformation. The converter imposes a check on the underlying table is a Flink's managed table and throws a `ValidationException` if it is a non-managed one. The converter also checks the partition spec and throws a `ValidationException` if the partition_spec is invalid.
  - Restrict `ALTER TABLE RESET` from removing the `connector` option key.
  

## Verifying this change

This change added tests and can be verified as follows:

  - *Added test in `FlinkSqlParserImplTest` to verify the syntax parsing*
  - *Added test in `SqlToOperationConverterTest` to verify the operation conversion*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (**yes** / no)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/ddl/AlterTableCompactOperation.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+    public AlterTableCompactOperation(
+            ObjectIdentifier tableIdentifier, @Nullable CatalogPartitionSpec partitionSpec) {
+        super(tableIdentifier);
+        this.partitionSpec = partitionSpec;
+    }
+
+    public Map<String, String> getPartitionSpec() {
+        return partitionSpec == null
+                ? new LinkedHashMap<>()
```
Comment (alexandru-tetelea) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/ddl/AlterTableCompactOperation.java:43]: Sorry, is it important to return a new empty instance or we can just return java.util.Collections.emptyMap()?

Comment (LadyForest) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/ddl/AlterTableCompactOperation.java:43]: You're right. We'd better return `Collections.emptyMap()`. It just hadn't hit me yet at that time.

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/operations/SqlToOperationConverter.java [hunk 2]
```diff
@@ -542,6 +551,60 @@ private Operation convertAlterTableReset(
...
     }
 
+    private Operation convertAlterTableCompact(
+            ObjectIdentifier tableIdentifier,
+            ResolvedCatalogTable resolvedCatalogTable,
+            SqlAlterTableCompact alterTableCompact) {
+        Catalog catalog =
+                catalogManager
+                        .getCatalog(tableIdentifier.getCatalogName())
+                        .orElseThrow(
```
Comment (JingsongLi) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/operations/SqlToOperationConverter.java:561]: orElse(null).
`isManagedTable` accepts nullable catalog.
We can throws exception to a unified mode in below.

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/operations/SqlToOperationConverter.java [hunk 3]
```diff
@@ -542,6 +551,60 @@ private Operation convertAlterTableReset(
...
+                                                        "Catalog %s not exists",
+                                                        tableIdentifier.getCatalogName())));
+        if (ManagedTableListener.isManagedTable(catalog, resolvedCatalogTable)) {
+            LinkedHashMap<String, String> partitionKVs = alterTableCompact.getPartitionKVs();
+            CatalogPartitionSpec partitionSpec = null;
+            if (partitionKVs != null) {
+                List<String> orderedPartitionKeys = resolvedCatalogTable.getPartitionKeys();
+                int index = 0;
+                String exMsg =
+                        orderedPartitionKeys.isEmpty()
```
Comment (JingsongLi) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/operations/SqlToOperationConverter.java:574]: I thought about it, maybe it doesn't have to be ordered, just hit the specified partition is OK.
We can just do something like `DynamicSinkUtils.validatePartitioning`

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/operations/SqlToOperationConverter.java [hunk 4]
```diff
@@ -542,6 +551,60 @@ private Operation convertAlterTableReset(
...
+                                                partitionKey, exMsg);
+                        throw new ValidationException(exMsg);
+                    }
+                    index++;
+                }
+                partitionSpec = new CatalogPartitionSpec(partitionKVs);
+            }
+            return new AlterTableCompactOperation(tableIdentifier, partitionSpec);
+        }
+        throw new UnsupportedOperationException(
```
Comment (JingsongLi) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/operations/SqlToOperationConverter.java:602]: `ValidationException`?

**Review by JingsongLi — CHANGES_REQUESTED** [review]
Thanks @LadyForest for your contribution, left some comments.
Can you separate to two commits? The first one is `Fix some error and change tests to junit5`, the second one is `Introduce ALTER.....`. In that way, can be more clear.

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/operations/SqlToOperationConverter.java (hunk 4 above)
Comment (LadyForest) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/operations/SqlToOperationConverter.java:602]: > `ValidationException`?

I'm fine with that.

**Comment by LadyForest** [comment]
Hi @JingsongLi, I've addressed your comments. Would you mind taking another look?

**Review by JingsongLi — APPROVED** [review]
Thanks @LadyForest for your update. Looks good to me!

**Pull request closed without merging**